Color classes

TakenPilot opened the pull request:
I don't know if you want this, but I figured I'd offer.  : )

This code does three things:
1) Removes the coffee-script, because hardly anyone can read it anymore and ES6 seems to be taking over.  Coffee-script also generates some strangely non-performant code.
2) Adds `colors` to the available options.  It allows users to specify their own ANSI color codes from 0 - 255 so they can implement theming, and handle the new background color in Windows 10 or light and dark themes on unix systems.
3) A major performance upgrade.  There was a `forEach` method that was looping over the input, and then looping over the tokens of that input again, which is O(n^n).  I removed that, and the unit tests still seem to pass.
4) I removed all the public-facing entry points that the user can't use, leaving only 'toHtml'.
5) Added eslint.

I needed this update for the performance and color codes for my own projects.  I'd love to get your input.

TakenPilot commented:
The build fails on node 4 and 5, but passes on 6.

rburns commented:
Hey, thanks for the patch. Here's my initial thoughts.

The change from coffeescript to vanilla js is probably a good thing. But, I'd like to support Node 4 and 6, while those versions are still being [maintained](https://github.com/nodejs/LTS).  What causes breakage in 4 and 5?

The performance improvements are quite welcome.

I'm on the fence about color handling I think. ie [classes vs palette manipulation](https://github.com/rburns/ansi-to-html/issues/5).  Part of that is it's not clear to me how the `colors` option [works](https://github.com/rburns/ansi-to-html/pull/33/commits/6f5c9f4231592f332b59990f15c72870abb085b9#diff-abe930d4e31ab573f519b475517e511cR304). It's a 256 element, 1 indexed, map between integers and hex-codes that gets merged with the default colors? If it's something like that, some tests that illustrate the merging or not merging behaviour might be good. Updating the [readme](https://github.com/TakenPilot/ansi-to-html/blob/8ba7264b97bb17eec0082a2433699f4e2ab3c371/README.md#options) would be good in any case.  

I'm not opposed to supporting both, I guess. I don't see any problems with them coexisting.

rburns commented:
Couple more notes. Cleaning up the public interface seems like a good thing. I think it was only incidentally public. 

It's not clear to me what's gained from [wrapping the constructor](https://github.com/TakenPilot/ansi-to-html/blob/color-classes/src/ansi_to_html.js#L415) in an auto-executing function.

TakenPilot commented:
The only thing that breaks 4 and 5 support is the use of the `let` keyword.  I'll can remove that.

For color support, it's exactly as you guessed.  I added a test to demonstrate, but I'll add a few more.  To explain quickly though, there is the default 256 color codes that map to the ANSI defaults.  If a user provides a map of colors as an option, those will be used instead.  Any color codes that are missing from their map will default back to the default ANSI color codes.

In most of the examples that people are providing in that thread, they only seem to care about the original 16 colors (red, blue, green, etc.), forgetting that there are actually 256 of them and that any of them might come from libraries that they're using or the programs generating the data that they're using.

Wrapping the constructor is a funny thing: V8 / Chrome / Mozilla will report any instances created as type `Filter` when debugging because of it.  We could give the class a better or more descriptive name though.

TakenPilot commented:
I replaced the `let` keyword with `var`, and removed the IIFE.  I also added some more unit tests to demonstrate overriding more specific colors.

Everything now passes in node 4 and 5 as well.  😁
